rust/core-lib/src/plugins/manifest.rs

```diff
@@ -107,10 +111,132 @@ pub enum PluginScope {
     SingleInvocation,
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize)]
+/// Represents a custom command provided by a plugin.
+pub struct Command {
+    /// Human readable title, for display in (for example) a menu.
+    pub title: String,
+    /// A short description of the command.
+    pub description: String,
+    /// Template of the command RPC as it should be sent to the plugin.
+    pub rpc_cmd: PlaceholderRpc,
+    /// A list of `CommandArgument`s, which the client should use to build the RPC.
+    pub args: Vec<CommandArgument>,
+}
+
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+/// A user provided argument to a plugin command.
+pub struct CommandArgument {
+    /// A human readable name for this argument, for use as placeholder
+    /// text or equivelant.
+    pub title: String,
+    /// A short (single sentence) description of this argument's use.
+    pub description: String,
+    pub key: String,
+    pub arg_type: ArgumentType,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    /// If `arg_type` is `Choice`, `options` must contain a list of options.
+    pub options: Option<Vec<ArgumentOption>>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub enum ArgumentType {
+    Number, Int, PosInt, Bool, String, Choice,
```

Comment:
I'll make this change, but I don't think it's a clear win (which I've felt about the majority of the feedback on this PR). In particular, I think that the type safety guarantees are overstated: these types should only ever be parsed at runtime (from manifest files or from json-rpc), and so the compile-time safety of a type system isn't particularly relevant, and I think that the ease of actually reading and writing the manifest files _is_ important, and this change doesn't help in either case.